lib/spack/spack/spec.py

```diff
@@ -1133,6 +1134,14 @@ def _add_default_platform(self):
         if arch and not arch.platform and (arch.os or arch.target):
             self._set_architecture(platform=spack.architecture.platform().name)
 
+    def _add_explicit_provider(self, virtual, spec):
+        # We cannot provide the same virtual multiple times
+        if virtual in self._explicit_providers:
+            msg = 'virtual dependency "{0}" cannot be specified multiple times'
+            raise ValueError(msg.format(virtual))
+
+        self._explicit_providers[virtual] = spec
```

Comment:
This is a fine check, (as it doesn't really know whether anything is virtual) but it *may* have to go away and be checked later.  It is right on the edge of including package/concretization semantics in the spec parser.

I'm on the fence because while I can't think of a *reason* we would ever want the same virtual twice in a DAG, I don't see why we shouldn't be able to *talk* about a DAG with that property.  A spec DAG *can* represent that -- it's just not a valid build.

Reply:
I completely agree with the reasoning above: one thing is construct an abstract spec, another is trying to concretize it using some constraints. I added the check above as I was trying to follow a practice that is already present for non-virtual dependencies:

https://github.com/spack/spack/blob/04c6a781b65450827f32cd6b6313bccaad3c725b/lib/spack/spack/spec.py#L1113-L1117

This results in the following behavior:
```console
Spack version 0.14.2
Python 3.7.4, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec('hdf5 ^zlib~shared ^zlib+shared')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 1035, in __init__
    spec_list = SpecParser(self).parse(spec_like)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/parse.py", line 152, in parse
    return self.do_parse()
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 4113, in do_parse
    self._parse_dependency(specs)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 4184, in _parse_dependency
    specs[-1]._add_dependency(dep, ())
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 1144, in _add_dependency
    "Cannot depend on '%s' twice" % spec)
spack.spec.DuplicateDependencyError: Cannot depend on 'zlib+shared' twice
```
Should we move both checks just before concretization and verify there the pre-requisites of an abstract spec?